File: 2.5 Histograma/histograma.py
#//////////////////////////////////////
#//Histograma
#//Autor: Bryan Medina
#/////////////////////////////////////


import cv2
import numpy as np
import matplotlib.pyplot as plt

#imagen de entrada
f = cv2.imread('zorro.jpg', 0)

def histograma(sourceImage):
    """calculo del histograma, la cual regresa el eje x con 256 valores de intensidad
    y la frecuencia de estos en la imagen, es decir, el histograma beibe"""

    #inicializacion del histograma y de los 256 de intensidad
    hist = np.zeros(256)
    x = np.array(range(256))

    #iteraremos sobre cada pixel de la imagen
    width, height = sourceImage.shape

    for i in range(width):
        for j in range(height):
            brillo = f[i, j]
            #aumentamos en 1 el nivel de brillo del pixel actual
            hist[brillo] += 1

    return x, hist


x, histF = histograma(f)

plt.figure()
plt.plot(x, histF)
plt.xlabel('niveles de intensidad')
plt.ylabel('número de pixeles')
plt.title('histograma')
plt.show()

# y buuuueeeno, tambien se puede usar la funcion de opencv y de numpy, ya cada quien
numpyHistogram, bins = np.histogram(f, 256, [0, 256])
opencvHistogram = cv2.calcHist([f], [0], None, [256], [0, 256])

plt.figure()
plt.plot(x, opencvHistogram)
plt.xlabel('niveles de intensidad')
plt.ylabel('número de pixeles')
plt.title('histograma')
plt.show()

File: 2.5 Histograma/histograma.m
%//////////////////////////////////////
%//Histograma
%//Autor: Bryan Medina
%/////////////////////////////////////

f = imread('zorro.jpg');
f = rgb2gray(f);

[m, n] = size(f);

hist = zeros(256, 1);
x = (1:256);

for i = 1:m
    for j = 1:n
        
        %aqui va el mas 1 por que los arrays empiezan en 1 en matlab, meh
        brillo = f(i, j) + 1 ;
        %se suma el pixel con este brillo en especifico, el histograma pues
        hist(brillo) = hist(brillo)+1;

    end
end

plot(x, hist);
xlabel('niveles de intensidad');
ylabel('numero de pixeles');
title('histograma');

% y pues matlab tambien tiene una funcion para calcular el histograma
matlabHist = imhist(f);
figure()
plot(x, matlabHist);
xlabel('niveles de intensidad');
ylabel('numero de pixeles');
title('histograma matlab');



